src/eval-frame/actions/actions.js

```diff
@@ -195,28 +318,36 @@ function evaluateCodeCell(cell) {
     if (sideEffectTarget) { sideEffectTarget.innerHTML = '' }
 
     dispatch(temporarilySaveRunningCellID(cell.id))
-    try {
-      output = window[languageModule][evaluator](code)
-    } catch (e) {
-      output = e
-      evalStatus = 'ERROR'
-    }
-    const updateCellAfterEvaluation = () => {
-      const cellProperties = { rendered: true }
-      if (evalStatus === 'ERROR') {
-        cellProperties.evalStatus = evalStatus
-      }
-      dispatch(updateCellProperties(cell.id, cellProperties))
-      // dispatch(incrementExecutionNumber())
-      dispatch(appendToEvalHistory(cell.id, cell.content, output))
-      dispatch(updateUserVariables())
-    }
 
-    const evaluation = Promise.resolve()
-      .then(updateCellAfterEvaluation)
-      .then(waitForExplicitContinuationStatusResolution)
-      .then(() => dispatch(temporarilySaveRunningCellID(undefined)))
-    return evaluation
+    ensureLanguageAvailable(cell.language, historyId, cell, state, dispatch)
```

Comment:
i'm not sure i'm following everything that is happening here, but i think we can simplify a lot of the stuff around how you're handling history items and passing around historyId, and i think this is the crux of it right here

this `ensureLanguageAvailable` block should go towards the top of evaluateCodeCell, before anything else is added to the eval history; if an eval is attemped on a cell with an unknown code type, _before_ the code/output block for the cell is added to the history, we pre-empt that with the console message that says "fetching resources to run python" (or whatevs), in a 'CELL_EVAL_INFO' history item, and then once that is done we can show the code and output in a "CELL_EVAL_VALUE" history item.

Reply:
If we do that (which was my initial pass), you end up with two console output blocks for a single cell -- one to load the language plugin and one for the output.  That didn't feel right to me.  We could add a "remove from history" action to remove the first one once it completes, but I thought it was simpler to just keep updating the same output cell, and changing the `historyType` accordingly.  But it's possible I'm missing a third option...

Reply:
ya that is the preferred behavior -- semantically, loading the plugin is _not_ part of the cell execution, it's an extra step that need to be taken when pyodide hasn't yet been loaded; what we're really doing is equivalent to automatically inserting a load plugin cell (which we will continue to support), and then evaling the code, and we should record those two steps as separate history items to show what is going on and maintain parallelism with the case of explicitly loading a language plugin. it should look something like:

```
============
Pyodide evaluator not yet available locally, fetching resource ...
(file download progress ...)
(initialization progress ...)
pyodide ready
============
>> [x**2 for x in range(5)]
------------
[1,4,9,16,25]
============
```

 i think this is a case of explicit>implicit: we want to be explicit about when we're inserting that extra step, and we want to leave the record of that step in console history to give our users more insight into what is happening behind the scenes.

Reply:
Ok.  I've updated this to do that.